src/services/providers/provider-sources/load-balance-provider.ts

```diff
@@ -0,0 +1,164 @@
+    if (transports.length === 0) throw new Error(`Chain with id ${chainId} not supported`);
+    return loadBalance(transports, this.config);
+  }
+}
+
+export function loadBalance(transports_: readonly Transport[], config: LoadBalanceConfig = {}): Transport {
+  const { key = 'load-balance', name = 'Load Balancing', minSuccessRate = 0.05, minSamples = 3, maxAttempts = 3, samplesTtl = '5m' } = config;
+
+  const rpcMetrics: MethodMetrics[] = transports_.map(() => ({}));
+
+  return ({ chain, timeout, ...rest }) => {
+    const transports = transports_.map((t) => t({ chain, timeout, ...rest }));
+
+    return createTransport({
+      key,
+      name,
+      async request({ method, params }) {
+        let availableTransports = transports.map((transport, index) => ({ transport, index }));
+
+        const processRequest = async (): Promise<any> => {
+          if (availableTransports.length === 0) {
+            throw new Error('All RPC attempts failed');
+          }
+
+          // Sort transports by score
+          availableTransports.sort((a, b) => {
+            const metricsA = rpcMetrics[a.index][method];
+            const metricsB = rpcMetrics[b.index][method];
+            if (!metricsA || !metricsB) return 0;
+            return calculateScore(metricsB) - calculateScore(metricsA);
+          });
+
+          const { transport, index } = availableTransports[0]; // Get the best transport
+
+          // Get the metrics for the method
+          let metrics = rpcMetrics[index][method];
+          const currentTime = Date.now();
+          if (!metrics || metrics.samples.length === 0 || currentTime - metrics.samples[0] > ms(samplesTtl)) {
+            metrics = initializeMetrics();
+            rpcMetrics[index][method] = metrics;
+          } else {
+            metrics = removeInvalidSamples(metrics, currentTime, samplesTtl);
+          }
+
+          if (metrics.samples.length >= minSamples && metrics.successRate < minSuccessRate) {
+            availableTransports.shift(); // Remove this transport. The best is the first, so we can shift
+            return processRequest(); // Try the next best transport
+          }
+
+          // Mark the request as pending and start the timer
+          metrics.pending++;
+          const start = currentTime;
+          try {
+            const response = await transport.request({ method, params });
+            rpcMetrics[index][method] = updateMetrics(start, metrics, true);
+            return response;
+          } catch (error) {
+            // If the error is a rejection, we should throw it
+            if (shouldThrow(error as Error)) throw error;
+
+            rpcMetrics[index][method] = updateMetrics(start, metrics, false);
+            availableTransports.shift(); // Remove this transport. The best is the first, so we can shift
+            return processRequest(); // Try the next best transport
+          }
+        };
+
+        return processRequest();
```

Comment:
I think that instead of creating a function and using recursion, we can simply have a `while(true)`. Since we will either return a response or throw an error